Modify Array/Dictionary::operator== to do real key/value comparison

touilleMan opened the pull request:
Here is my attempt to fix the current infamous dictionary comparison behavior (see #27615)

Basically the issue is currently two Dictionary object only considered equal if they point on the same underlying hashmap.
This is very error prone given this behavior leaks into GDScript:
```gdscript
var a = {}
var b = {}
var c = a
print(a == b)  # false
print(a == c)  # true
```

There is already PR #29222 that tries to fix this by modifying Variant comparison. However I believe this is not the right way to go given it only fix GDScript, but not GDNative !

Currently GDNative's `godot_dictionary_operator_equal` is pretty useless (I would even say it is extremely error prone, I've personally been bitten multiple time by this function) and so GDNative users has to implement their own comparison (see https://github.com/touilleMan/godot-python/blob/72df9d1b38120aeec3c702004e14dcb82e95bde5/tools/builtins_templates/dictionary.tmpl.pxi#L180-L185) which is both cumbersome and inefficient.

Given I couldn't really figure out the usecase for this underlying-hashmap-only comparison, I've recompiled a modified version of Godot with `Dictionary::operator==` and `Dictionary::operator!=` removed (see https://github.com/touilleMan/godot/commit/2409aa62bed81be948e3efbc99ae3c578fda6502).
The only place those operator are used are in `variant_op.cpp` in the `OP_EQUAL` and `OP_NOT_EQUAL` operators.
So I guess I can officially say the current implementation of `Dictionary::operator==` is useless \o/

Of course, I think we should do the same for `Array::operator==`, I'll update this PR with those change if we agree on the idea ;-)

*Bugsquad edit:*
- Fixes #25375.
- Fixes #27615.
- Fixes #29221.

touilleMan committed:
Modify Dictionary::operator== to do real key/value comparison with recursive support (and add unittests)

pouleyKetchoupp commented:
From what you discovered about the use of == operators in the code base, it does seem like this is a better solution than PR #29222.

Two things I would like to mention:

- In that PR I've added a fix in project_settings_editor.cpp because deep dictionary comparison was causing a regression:
https://github.com/godotengine/godot/pull/29222#issuecomment-526848775

- Reduz had mentioned a potential issue with cyclical inclusions, which I figured is probably the case where the same dictionary is added inside of itself (for the most simple case), and it would cause an infinite loop.
https://github.com/godotengine/godot/pull/29222#issuecomment-503520898

touilleMan commented:
@pouleyKetchoupp thanks a lot for you inputs ;-)

> In that PR I've added a fix in project_settings_editor.cpp because deep dictionary comparison was causing a regression:

I'll had to this PR your change related to this issue then :+1: 

> Reduz had mentioned a potential issue with cyclical inclusions, which I figured is probably the case where the same dictionary is added inside of itself (for the most simple case), and it would cause an infinite loop.

That's a really good point !
In fact recursion is already a trouble in the current code base, the following code cause a segfault on Godot 3.2:
```gdscript
var recursive_a_1 = [1]
recursive_a_1.append(recursive_a_1)
print("recursive_a_1 == recursive_a_1 ==> ", recursive_a_1 == recursive_a_1)
```

This is interesting given doing `print("recursive: ", recursive_a_1)` works as expected (it output `recursive: [1, [...]]`). This works because `Variant::stringify` keeps `List<const void *> stack` to track the recursivity.

So I've modified the PR to handle recursive check with a similar stack system.

However I fell like this code would benefit a lot from unit testing... but there is no such thing in Godot (well there is the https://github.com/godotengine/godot-tests repo, but it doesn't seems really used)

Here is the tests I've done manually:
```gdscript
var x1 = {}
var y1 = {}
x1[1] = y1
y1[1] = y1

var x2 = {}
var y2 = {}
x2[1] = y2
y2[1] = y2

print("x1 == x2 ==> ", x1 == x2)
# Test dictionary

var d1 = {1: 1}
var d2 = {1: 1}
var d1_ref2 = d1
var other_d = {1: 2}

print("d1 == d2 ==> ", d1 == d2)
print("d1 == d1_ref2 ==> ", d1 == d1_ref2)
print("d1 != other_d ==> ", d1 != other_d)

var nested1_d1 = {1: d1, 2: 2}
var nested2_d1 = {1: d1, 2: 2}
var nested3_d2 = {1: d2, 2: 2}
var other_d_nested = {1: other_d, 2: 2}

print("nested1_d1 == nested2_d1 ==> ", nested1_d1 == nested2_d1)
print("nested1_d1 == nested3_d2 ==> ", nested1_d1 == nested3_d2)
print("nested1_d1 != other_d_nested ==> ", nested1_d1 != other_d_nested)

var recursive_d_1 = {1: 1}
recursive_d_1[2] = recursive_d_1	
print("recursive_d_1 == recursive_d_1 ==> ", recursive_d_1 == recursive_d_1)

var recursive_d_2 = {1: 1}
recursive_d_2[2] = recursive_d_2
print("recursive_d_1 == recursive_d_2 ==> ", recursive_d_1 == recursive_d_2)

var cross_recursive_d_1 = {1: 1}
var cross_recursive_d_2 = {1: 1}
cross_recursive_d_1[2] = cross_recursive_d_2
cross_recursive_d_2[2] = cross_recursive_d_1
print("cross_recursive_d_1 == cross_recursive_d_2 ==> ", cross_recursive_d_1 == cross_recursive_d_2)

if false:
	var key_recursive_d_1 = {1: 1}
	var key_recursive_d_2 = {1: 1}
	# BUG !!!! This Crash here !
	key_recursive_d_1[key_recursive_d_2] = 1
	key_recursive_d_2[key_recursive_d_1] = 1
	print("key_recursive_d_1 == key_recursive_d_2 ==> ", key_recursive_d_1 == key_recursive_d_2)

if false:
	var x = ['x']
	var y = ['y']
	var z = ['z']
	y.append(y)
	x.append(y)
	x.append(y)
	# BUG !!!!! this prints "[x, [y, [...]], [...]]" but it should be "[x, [y, [...]], [y, [...]]]"
	print(x)

# Test array

var a1 = [1]
var a2 = [1]
var a1_ref2 = a1
var other_a = [2]

print("a1 == a2 ==> ", a1 == a2)
print("a1 == a1_ref2 ==> ", a1 == a1_ref2)
print("a1 != other_a ==> ", a1 != other_a)

var nested1_a1 = [a1, 2]
var nested2_a1 = [a1, 2]
var nested3_a2 = [a2, 2]
var other_a_nested = [other_a, 2]

print("nested1_a1 == nested2_a1 ==> ", nested1_a1 == nested2_a1)
print("nested1_a1 == nested3_a2 ==> ", nested1_a1 == nested3_a2)
print("nested1_a1 != other_a_nested ==> ", nested1_a1 != other_a_nested)

var recursive_a_1 = [1]
recursive_a_1.append(recursive_a_1)
print("recursive: ", recursive_a_1)
print("recursive_a_1 == recursive_a_1 ==> ", recursive_a_1 == recursive_a_1)

var recursive_a_2 = [1]
recursive_a_2.append(recursive_a_2)
print("recursive_a_1 == recursive_a_2 ==> ", recursive_a_1 == recursive_a_2)

var cross_recursive_a_1 = [1]
var cross_recursive_a_2 = [1]
cross_recursive_a_1.append(cross_recursive_a_2)
cross_recursive_a_2.append(cross_recursive_a_1)
print("cross_recursive_a_1 == cross_recursive_a_2 ==> ", cross_recursive_a_1 == cross_recursive_a_2)
```

As you can see I've also discovered two unrelated bugs ;-)

RandomShaper commented:
What's the status of this?
I haven't reviewed the code thoroughly, but the change is very welcome.

touilleMan commented:
@RandomShaper It should be ready to merge, my only concern is there is currently no place to add unittests in godot :'(

Calinou commented:
> It should be ready to merge, my only concern is there is currently no place to add unittests in godot :'(

See https://github.com/godotengine/godot/pull/30743.

touilleMan commented:
I've rebased this PR ;-)

touilleMan commented:
@akien-mga sorry to directly ping you, but I'm not sure what can be done to get this and #35813 merged...

pouleyKetchoupp commented:
Just tested on master.
This PR fixes #29221 and possibly some other issues related to failing dictionary comparison.

fire commented:
> This branch has conflicts that must be resolved

Can this be rebased?

touilleMan commented:
@fire I've rebased the PR

However it seems to have an issue related to `variant_call.cpp` when running the binary:
```
$  .\bin\godot.windows.opt.tools.64.exe
ERROR: Wrong argument name count supplied for method:  Dictionary::hash
   at: (core\variant_call.cpp:571)
ERROR: Wrong argument name count supplied for method:  Array::hash
   at: (core\variant_call.cpp:571)
```

I suspect this is linked to what @reduz said in #42780:
> Note: while it should build ok this PR breaks release target (I think its broken already..), a subsequent PR refactoring method binds should fix it.

Zylann commented:
Hang on, this basically replaces the behavior of `==` to compare by contents instead of reference? Or am I misunderstanding?
Because when this happened for arrays, we ended up with the impossibility to compare by ref, leading to https://github.com/godotengine/godot/issues/33627
(see also discussion on https://github.com/godotengine/godot/issues/27615#issuecomment-681961385)
Maybe it makes sense if you come from Python, but every other languages I used does not do that, and instead offers specific ways of comparing by value, instead of completely replacing it.

touilleMan commented:
> Hang on, this basically replaces the behavior of == to compare by contents instead of reference?

The idea is to replace the behavior of` `==` in GDScript and in GDNative `xxx_operator_equal`, the simplest way to do that was to rewrite the C++'s `operator==` for Array and Dictionary.
It also seems a good idea to have a coherent comparison model: either GDScript `==`, GDNative `xxx_operator_equal` and C++ `operator==` all do pointer comparison, either they all do value comparison.
Having a mix of that seems seems like a recipe for nasty bug in my opinion ;-)

>  Because when this happened for arrays, we ended up with the impossibility to compare by ref, leading to #33627 and personally I didn't like that when it was introduced.

`Array.id()`/`Dictionary.id()` is pretty close to your need (the only trouble is id() returns NULL when the array/dictionary is empty, so you cannot directly do `is_different = a.id() != b.id()`)
I guess we could expose this `id()` method in GDScript (or a sightly modified version to overcome the NULL issue).

>  Maybe it makes sense if you come from Python, but every other languages I used does not do that, and at least offers the option to compare either by ref or by value, instead of completely replacing it.

In Python you can do `a is b` to test if `a` and `b` point to the same underlying object.
The point I guess is what's the least surprising behavior for `==` for GDScript users.
Given GDScript is inspired by Python I would say the `==` as value comparison is the most logical behavior.
But I think the most important thing is to have both kind of comparisons (per value and per pointer) available and documented (so that you )

touilleMan commented:
It took me the entire weekend, but I think we're basically done here ^^

I've added plenty of unittests, which in turn allowed me to discover&fix a good amount of recursion related bugs (typically in `duplicate()` and in the variant writer)

The is still some minor things to adjust:
- the recursive comparison function is called `hash_compare` in Variant and `recursive_equal` in Array/Dictionary. I find `recursive_equal` more logical given we also have `recursive_hash` and `recursive_duplicate` (each being called by the regular `operator==`/`hash`/`duplicate` method starting the recursive algo)
- GDNative expose `godot_array/dictionary_hash_compare` which are no longer needed and should be removed
- as @Zylann remarked, a by ref comparison would be welcome, I propose [this](https://github.com/godotengine/godot-proposals/issues/874#issuecomment-711139350) (but this is something for another PR)
- dictionary using array or dictionary as keys cause recursion and comparison issues (see the commented tests).  It's a complex topic given the hashmap used internally by Dictionary doesn't support recursion limit. I'm not sure what to do there (but anyway it's a very specific topic, so I think is ok to merge this PR first, then wonder what to do ^^)

@godotengine/pr-mergers sorry to tag you directly, but this PR is 8 months old now and still didn't get any proper feedback. Given it touches multiples core files I fear it will end up with multiple of conflicts very quickly 😃

jonbonazza commented:
Im still not totally on board with this change due to reasons already discussed, anf adding the reason of performance. Comparing large arrays and dicts cam be very slow with this change and imo comparing an entire dict or array for value equality in the first place is often a sign of code smell. It's simply not a very useful thing to do in general. If it's being attempted, it usually means the algorithm in question isnt ideal and can be improved.

jordo commented:
I'll add my comment here as a way to support this PR big time.

But anecdotally, I just got burned working in c++ with this.  It's really non-intuitive to expose the == operator in the data structures (Dictionary/Array) header, and not have it operate as a value based comparator.

I've had some discussion on IRC and I'm in support of this.  Anything to align with whats intuitive is a good thing.

Consider what developers are used to:

```
std::map<int,int> a; 
a[1] = 1; 
std::map<int,int> b; 
b[1] = 1;  
bool is_equal = a == b;
```

`is_equal` is clearly true here, the ==operator isn't going to actually compare the backing pointers of the two maps..  Actually the fact that the == operator is even implemented IMPLIES that it should be a value based comparison! 

I really don't understand any opposition to this.  This is pretty standard IMO.

+1 for this PR.  Helps me immensely, and I won't have to implement it myself!

Zylann commented:
@jordo your example comes from C++, which is very far and different from a regular scripting language. C++ even gives you plenty of alternatives. While GDScript gives you none.

jordo commented:
Yes my example is c++ because the engine is written in c++.  The godot codebase should be consistent and approachable for c++ developers.  

If that's the desired functionality in GDScript, or whatever other scripting language, than that's perfectly fine.  Map the evaluate() and OP_EQUAL scripting operator to some other private or other function in dictionary that just compares the backing reference.

But in c++ ```bool operator==(const Dictionary &p_dictionary) const;``` implies that func is a value based comparator, implementing == through the containers key's and values.

reduz commented:
It would be good to merge this for 4.0, although an equality with a deep comparison by default is likely better. IMO we should consider equality of objects with recursive referencing an error so I would put a high max recursion value by default (like 20k, configurable in project settings) and if it reaches it should error. I think in practice given we compare pointers, reaching this situation should be extremely rare.

aaronfranke commented:
@touilleMan This PR needs to be rebased on the latest master when you have the time.

akien-mga commented:
The doc changes in the CI build seem to be due to changes in the printing of arrays:
https://github.com/godotengine/godot/pull/35816/checks?check_run_id=3442432599
https://github.com/godotengine/godot/pull/35816/files#diff-8fa43a0158f72fe267697c83f42d65e567c538c003320e7a9417faffd51f8bfeR1648

Might be a rebase mistake as those changed recently IINM.

touilleMan commented:
@akien-mga Thanks for the tip ! I'll have a look at this ;-)

touilleMan committed:
Correct CodeEdit documentation mentioning parent property due to incorrect array comparison

touilleMan commented:
@akien-mga  the last error in the build involve the documentation:

```diff
iff --git a/doc/classes/CodeEdit.xml b/doc/classes/CodeEdit.xml
index 93f72d45ae..60efd5d3a5 100644
--- a/doc/classes/CodeEdit.xml
+++ b/doc/classes/CodeEdit.xml
@@ -494,7 +494,6 @@
                <member name="line_length_guidelines" type="int[]" setter="set_line_length_guidelines" getter="get_line_length_guidelines" default="[]">
                        Draws vertical lines at the provided columns. The first entry is considered a main hard guideline and is draw more prominently
                </member>
-               <member name="structured_text_bidi_override_options" type="Array" setter="set_structured_text_bidi_override_options" getter="get_structured_text_bidi_override_options" override="true" default="[]" />
                <member name="symbol_lookup_on_click" type="bool" setter="set_symbol_lookup_on_click_enabled" getter="is_symbol_lookup_on_click_enabled" default="false">
                        Set when a validated word from [signal symbol_validate] is clicked, the [signal symbol_lookup] should be emitted.
                </member>
```

It is due to new array comparison behavior in here:
https://github.com/godotengine/godot/blob/84faf3954417c0ee58df79c0d2cc09309825878a/editor/doc_tools.cpp#L309

Here, when dealing with `CodeEdit::structured_text_bidi_override_options`, `default_value` and `base_default_value` are both an empty array. However before this PR the `default_value == base_default_value` would still be false (given comparison was done on pointers and not on values).

My guess is the per-pointer comparison was not intended here given this part of the code aims at not duplicating parent attributes that have been left untouched by children.

So my last commit 2351380d74f7604a9707f8e7b5892cec10dbb929 corrects the documentation, this PR should be finally ready to merge now ;-)

RandomShaper reviewed (changes requested):
I guess the commits could benefit from some squashing.

touilleMan commented:
@RandomShaper rebased&squashed, ready for merge ;-)

akien-mga commented:
The documentation for Dictionary (and maybe Array) should be updated to reflect these changes. At least Dictionary's description still has an explicit warning that Dictionary comparisons are not based on contents.

IIUC, we now lack a way to compare Arrays and Dictionaries by reference. It would be good to add one so that we can close all related issues / proposals as solved.
